pkg/controllers/federatedhpa/util.go

```diff
@@ -0,0 +1,507 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package federatedhpa
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+
+	fedcorev1a1 "github.com/kubewharf/kubeadmiral/pkg/apis/core/v1alpha1"
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/common"
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/follower"
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/scheduler"
+	"github.com/kubewharf/kubeadmiral/pkg/util/fedobjectadapters"
+	"github.com/kubewharf/kubeadmiral/pkg/util/pendingcontrollers"
+	"github.com/kubewharf/kubeadmiral/pkg/util/worker"
+)
+
+const (
+	PropagationPolicyKind        = "PropagationPolicy"
+	ClusterPropagationPolicyKind = "ClusterPropagationPolicy"
+
+	HPAScaleTargetRefPath = "hpa.kubeadmiral.io/scale-target-ref-path"
+	FedHPANotWorkReason   = "hpa.kubeadmiral.io/fed-hpa-not-work-reason"
+
+	HPAEnableKey = "fed-hpa-enabled"
+
+	EventReasonUpdateHPASourceObject = "UpdateHPASourceObject"
+	EventReasonUpdateHPAFedObject    = "UpdateHPAFedObject"
+)
+
+type Resource struct {
+	gvk       schema.GroupVersionKind
+	namespace string
+	name      string
+}
+
+func (r Resource) QualifiedName() common.QualifiedName {
+	return common.QualifiedName{
+		Namespace: r.namespace,
+		Name:      r.name,
+	}
+}
+
+func fedObjectToSourceObjectResource(object metav1.Object) (Resource, error) {
+	fedObj, _ := object.(fedcorev1a1.GenericFederatedObject)
+	metadata, err := fedObj.GetSpec().GetTemplateMetadata()
+	if err != nil {
+		return Resource{}, err
+	}
+	return Resource{
+		name:      metadata.GetName(),
+		namespace: metadata.GetNamespace(),
+		gvk:       metadata.GroupVersionKind(),
+	}, nil
+}
+
+func policyObjectToResource(object metav1.Object) Resource {
+	if cpp, ok := object.(*fedcorev1a1.ClusterPropagationPolicy); ok {
+		return Resource{
+			name:      cpp.GetName(),
+			namespace: cpp.GetNamespace(),
+			gvk:       cpp.GroupVersionKind(),
+		}
+	}
+
+	if pp, ok := object.(*fedcorev1a1.PropagationPolicy); ok {
+		return Resource{
+			name:      pp.GetName(),
+			namespace: pp.GetNamespace(),
+			gvk:       pp.GroupVersionKind(),
+		}
+	}
+	return Resource{}
+}
+
+func generateFederationHPANotWorkReason(
```

Comment:
can we add pp name and kind into reason? it will be useful for users to know what to do next

Reply:
Yes, we can. The code has been updated.